### PR TITLE
Add more Panel/ScrollableControl tests and fix a bug

### PR DESCRIPTION
**Pull request by hughbe**
Contributes to #817

### CODE REVIEW

File: src/System.Windows.Forms/src/System/Windows/Forms/Control.cs [hunk 1]
```diff
@@ -935,7 +935,7 @@ private AmbientProperties AmbientPropertiesService
                 {
                     if (Site != null)
                     {
-                        props = (AmbientProperties)Site.GetService(typeof(AmbientProperties));
+                        props = Site.GetService(typeof(AmbientProperties)) as AmbientProperties;
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/Control.cs]: I don't agree with this.
If I ask to get an `AmbientProperties` service and `GetService()` call returns me something else - it is a bug in `GetService()` implementation, and we must fail.

/cc: @Tanya-Solyanik

Comment (hughbe) [reply src/System.Windows.Forms/src/System/Windows/Forms/Control.cs]: This is #817

**Comment by Tanya-Solyanik** [comment]
@hughbe  - could you please separate addition of unit tests from the GetService change? It would be easier to track.

**Comment by hughbe** [comment]
Sure! I've removed the `Control.Site` fixes and tests from this PR :)

**Comment by hughbe** [comment]
@RussKie good to merge ?

File: src/System.Windows.Forms/src/System/Windows/Forms/ScrollableControl.cs [hunk 2]
```diff
@@ -130,51 +107,36 @@ public virtual bool AutoScroll
...
                     throw new ArgumentOutOfRangeException(nameof(value), value, string.Format(SR.InvalidArgument, nameof(AutoScrollMargin), value));
                 }
+
                 SetAutoScrollMargin(value.Width, value.Height);
             }
         }
 
-        /// <summary>
-        ///    <para>Gets or sets the location of the auto-scroll position.</para>
-        /// </summary>
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/ScrollableControl.cs]: Please reinstate the doc

File: src/System.Windows.Forms/src/System/Windows/Forms/ScrollableControl.cs [hunk 3]
```diff
@@ -294,121 +243,52 @@ Rectangle IArrangedElement.DisplayRectangle
...
-            }
-            set
-            {
-                SetScrollState(ScrollStateHScrollVisible, value);
-            }
-        }
-
-        /// <summary>
-        ///    <para>Gets the Horizontal Scroll bar for this ScrollableControl.</para>
-        /// </summary>
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/ScrollableControl.cs]: ditto

File: src/System.Windows.Forms/src/System/Windows/Forms/ScrollableControl.cs [hunk 4]
```diff
@@ -294,121 +243,52 @@ Rectangle IArrangedElement.DisplayRectangle
...
 
+        [SRCategory(nameof(SR.CatLayout))]
+        [SRDescription(nameof(SR.ScrollableControlVerticalScrollDescr))]
+        [Browsable(false)]
+        [EditorBrowsable(EditorBrowsableState.Always)]
+        public VScrollProperties VerticalScroll => verticalScroll ??= new VScrollProperties(this);
 
         /// <summary>
-        ///    <para>Gets the Veritcal Scroll bar for this ScrollableControl.</para>
-        /// </summary>
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/ScrollableControl.cs]: ditto

File: src/System.Windows.Forms/src/System/Windows/Forms/ScrollableControl.cs [hunk 5]
```diff
@@ -862,17 +712,14 @@ protected override void ScaleCore(float dx, float dy)
...
         /// </summary>
         protected override void ScaleControl(SizeF factor, BoundsSpecified specified)
         {
             ScaleDockPadding(factor.Width, factor.Height);
             base.ScaleControl(factor, specified);
         }
 
-        // internal for the respective scroll bars to call so that the Display Rectangle is set to 
-        // enable the visual scroll effect.
-        //
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/ScrollableControl.cs]: It would be good to reinstate the comment in a modified form, e.g. `Allows to set the <see cref="DisplayRectangle" /> to enable the visual scroll effect.`

File: src/System.Windows.Forms/src/System/Windows/Forms/Layout/DockAndAnchorLayout.cs [hunk 6]
```diff
@@ -779,7 +779,6 @@ public static void SetAnchor(IArrangedElement container, IArrangedElement elemen
                     LayoutTransaction.DoLayout(element.Container, element, PropertyNames.Anchor);
                 }
             }
-            Debug.Assert(GetAnchor(element) == value, "Error setting Anchor value.");
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/Layout/DockAndAnchorLayout.cs]: Why is this deleted?
We had a discussion about this recently and the direction was not to delete these.
If it is causing issue in unit tests, please use `TraceListenerlessContext`

Comment (hughbe) [reply src/System.Windows.Forms/src/System/Windows/Forms/Layout/DockAndAnchorLayout.cs]: If you pass an invalid anchor value then it gets normalized to only include `Top/Left/Right/Bottom`:
```cs
public static IEnumerable<object[]> Anchor_Set_TestData()
{
    yield return new object[] { (AnchorStyles)(-1), AnchorStyles.Top | AnchorStyles.Bottom | AnchorStyles.Left | AnchorStyles.Right };
    yield return new object[] { (AnchorStyles)int.MaxValue, AnchorStyles.Top | AnchorStyles.Bottom | AnchorStyles.Left | AnchorStyles.Right };
}

[Theory]
[MemberData(nameof(Anchor_Set_TestData))]
public void Control_Anchor_Set_GetReturnsExpected(AnchorStyles value, AnchorStyles expected)
{
    var control = new Control
    {
        Anchor = value
    };
    Assert.Equal(expected, control.Anchor);
    Assert.Equal(DockStyle.None, control.Dock);

    // Set same.
    control.Anchor = value;
    Assert.Equal(expected, control.Anchor);
    Assert.Equal(DockStyle.None, control.Dock);
}
```

I'm very much against keeping this debug assert that is essentially a unit test inside product code. I'm also against using `TraceListenerlessContext` as this has little function

**Comment by RussKie** [comment]
@hughbe I am having troubles correlating the PR subject to the actual changes.
I see changes to `Panel` and `ScrollableControl` and related tests.

I would appreciate it if you could clarify what the bug was. Also why `DockAndAnchorLayout` changed? 
Extra tests are great, but why are they done in the PR?

Thank you

**Comment by RussKie** [comment]
Looks like there is a conflict.

File: src/System.Windows.Forms/src/System/Windows/Forms/Panel.cs [hunk 7]
```diff
@@ -274,20 +238,13 @@ internal override void PrintToMetaFileRecursive(HandleRef hDC, IntPtr lParam, Re
...
-
         /// <summary>
-        ///  Returns a string representation for this control.
+        /// Returns a string representation for this control.
         /// </summary>
         public override string ToString()
         {
             string s = base.ToString();
-            return s + ", BorderStyle: " + StringFromBorderStyle(borderStyle);
+            return s + ", BorderStyle: " + typeof(BorderStyle).ToString() + "." + _borderStyle.ToString();
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/Panel.cs]: This could be slightly improved by moving the static part into a static variable, e.g. `private static readonly string str = ", BorderStyle: " + typeof(BorderStyle) + ".";` (called `Const` in the branchmark):
         
image
